src/main/java/org/folio/marc4ld/service/marc2ld/preprocessor/impl/TitlePreprocessor.java

```diff
@@ -0,0 +1,80 @@
+package org.folio.marc4ld.service.marc2ld.preprocessor.impl;
+
+import static org.folio.marc4ld.util.Constants.A;
+import static org.folio.marc4ld.util.Constants.B;
+import static org.folio.marc4ld.util.Constants.S;
+import static org.folio.marc4ld.util.Constants.T;
+import static org.folio.marc4ld.util.Constants.TAG_245;
+import static org.folio.marc4ld.util.Constants.TAG_776;
+import static org.folio.marc4ld.util.MarcUtil.getSubfieldValue;
+import static org.folio.marc4ld.util.MarcUtil.isSubfieldPresent;
+
+import java.util.List;
+import java.util.Optional;
+import lombok.RequiredArgsConstructor;
+import org.folio.marc4ld.service.marc2ld.preprocessor.DataFieldPreprocessor;
+import org.marc4j.marc.DataField;
+import org.marc4j.marc.MarcFactory;
+import org.marc4j.marc.Record;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class TitlePreprocessor implements DataFieldPreprocessor {
+
+  private static final List<String> TAGS = List.of(TAG_776);
+
+  private final MarcFactory marcFactory;
+
+  @Override
+  public Optional<DataField> preprocess(PreprocessorContext context) {
+    var originalDataField = context.dataField();
+    var preprocessedDataField = duplicateDataField(originalDataField);
+    addSubfieldsRelatedToTitle(context, originalDataField, preprocessedDataField);
+    return Optional.of(preprocessedDataField);
+  }
+
+  @Override
+  public List<String> getTags() {
+    return TAGS;
+  }
+
+  private DataField duplicateDataField(DataField originalDataField) {
+    var dataField = marcFactory.newDataField(originalDataField.getTag(), originalDataField.getIndicator1(),
+      originalDataField.getIndicator2());
+    originalDataField.getSubfields()
+      .stream()
+      .filter(subfield -> S != subfield.getCode() || T != subfield.getCode())
+      .forEach(dataField::addSubfield);
+    return dataField;
+  }
+
+  private void addSubfieldsRelatedToTitle(PreprocessorContext context, DataField originalDataField,
+                                          DataField preprocessedDataField) {
+    String workTitle;
+    String instanceTitle;
+    if (isSubfieldPresent(S, originalDataField)) {
+      workTitle = getSubfieldValue(S, originalDataField);
+      instanceTitle = isSubfieldPresent(T, originalDataField)
+        ? getSubfieldValue(T, originalDataField)
+        : getTitle(context.marcRecord());
+    } else if (isSubfieldPresent(T, originalDataField)) {
+      workTitle = getSubfieldValue(T, originalDataField);
+      instanceTitle = getSubfieldValue(T, originalDataField);
+    } else {
+      workTitle = getTitle(context.marcRecord());
+      instanceTitle = getTitle(context.marcRecord());
+    }
+    preprocessedDataField.addSubfield(marcFactory.newSubfield(S, workTitle));
```

Comment:
I think this logic will result in creating two` $s`s if `preprocessedDataField` already had a `$s`
For example, let us assume the original MARC record was `776 $s some title`. After executing this method, the MARC becomes `776 $s some title $s some title`

Reply:
It shouldn't because of this line in ```duplicateDataField``` method
```java
.filter(subfield -> S != subfield.getCode() || T != subfield.getCode())
```
Anyway, your option looks much cleaner, so I will apply it. Thank you!

Reply:
Done